crates/bevy_ecs/src/world/mod.rs

```diff
@@ -439,21 +464,105 @@ impl World {
     /// let e1 = world.spawn().insert(A).id();
     /// let e2 = world.spawn().insert_bundle((A, B)).id();
     ///
-    /// let mut query = world.query_filtered::<Entity, With<B>>();
-    /// let matching_entities = query.iter(&world).collect::<Vec<Entity>>();
+    /// let mut query = world.query_state_filtered::<Entity, With<B>>();
+    /// let system_counter = world.get_exclusive_system_counter();
+    /// let global_system_counter = world.get_global_system_counter();
+    /// let matching_entities = query.iter(&world, system_counter, global_system_counter).collect::<Vec<Entity>>();
     ///
     /// assert_eq!(matching_entities, vec![e2]);
     /// ```
     #[inline]
-    pub fn query_filtered<Q: WorldQuery, F: WorldQuery>(&mut self) -> QueryState<Q, F>
+    pub fn query_state_filtered<Q: WorldQuery, F: WorldQuery>(&mut self) -> QueryState<Q, F>
     where
         F::Fetch: FilterFetch,
     {
         QueryState::new(self)
     }
 
-    /// Returns an iterator of entities that had components of type `T` removed since the last call
-    /// to [World::clear_trackers].
+    /// Returns a [DirectQuery] for the given [WorldQuery],
+    /// which is used to directly query the [World],
+    /// usually in exclusive systems.
+    /// ```
+    ///
+    /// use bevy_ecs::{entity::Entity, world::World};
+    ///
+    /// #[derive(Debug, PartialEq)]
+    /// struct Position {
+    ///   x: f32,
+    ///   y: f32,
+    /// }
+    ///
+    /// struct Velocity {
+    ///   x: f32,
+    ///   y: f32,
+    /// }
+    ///
+    /// let mut world = World::new();
+    /// let entities = world.spawn_batch(vec![
+    ///     (Position { x: 0.0, y: 0.0}, Velocity { x: 1.0, y: 0.0 }),    
+    ///     (Position { x: 0.0, y: 0.0}, Velocity { x: 0.0, y: 1.0 }),    
+    /// ]).collect::<Vec<Entity>>();
+    ///
+    /// let mut query = world.query::<(&mut Position, &Velocity)>();
+    /// for (mut position, velocity) in query.iter_mut() {
+    ///    position.x += velocity.x;
+    ///    position.y += velocity.y;
+    /// }     
+    ///
+    /// assert_eq!(world.get::<Position>(entities[0]).unwrap(), &Position { x: 1.0, y: 0.0 });
+    /// assert_eq!(world.get::<Position>(entities[1]).unwrap(), &Position { x: 0.0, y: 1.0 });
+    /// ```
+    #[inline]
+    pub fn query<Q: WorldQuery>(&mut self) -> DirectQuery<Q, ()> {
+        let state = QueryState::new(self);
+        let global_system_counter = self.get_global_system_counter_unordered();
+        unsafe {
+            DirectQuery::new(
+                self,
+                state,
+                self.get_exclusive_system_counter(),
```

Comment:
Will this actually behave as expected when it isn't running inside an exclusive system (ex: someone is interacting with World directly, either because they don't want to use a Schedule or because they are using something like AppBuilder).

Reply:
Exclusive systems appear to be the only things setting World's exclusive system counter.

Reply:
Anyone not using systems would need to call `world.increment_global_system_counter()` whenever appropriate if they want to detect changes, like we do in the tests.
Using systems without a schedule works, since normal and exclusive systems call `world.increment_global_system_counter()` when they are run (see `FunctionSystem::run_unsafe()`).
Changing something in the `AppBuilder`'s world before running the schedule will work as expected because the global system counter is initialized to `0` and the systems counters are initialized at `u32::MAX` (i.e. changes will be detected on the first run of the systems).

Reply:
I misread the question and went on a tangent, sorry. It works outside of exclusive systems because `world.exclusive_system_counter` is initialized to `u32::MAX`, so a direct query will detect any change since the creation of the world (see the comment in the `Default` impl for `World`).

Reply:
No worries. I missed the manual assignment to world.exclusive_system_counter in the tests. We might want to build an abstraction that makes it easier to "clear" changes directly on world (like clear_trackers).

Reply:
The only use-case I see for that is in bevy's own tests, so I think we should avoid cluttering the API for it. These two operations are usually not done together, it's only the case here because tests are "pretending" to be one-exclusive-system schedules.